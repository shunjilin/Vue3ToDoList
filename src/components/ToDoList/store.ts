import { inject, InjectionKey, provide, reactive, toRefs, readonly } from 'vue';

interface ToDoItem {
  completed: boolean;
  title: string;
  created: Date;
}

interface State {
  list: ToDoItem[];
  addToDoInput: string;
}

export function toDoListStore() {
  const state = reactive<State>({
    list: [],
    addToDoInput: '',
  });

  function addToDo() {
    if (state.addToDoInput) {
      state.list = [
        ...state.list,
        {
          title: state.addToDoInput,
          completed: false,
          created: new Date(Date.now()),
        },
      ];
      state.addToDoInput = '';
    }
  }

  function removeToDo(indexToRemove: number) {
    state.list = state.list.filter((todo, index) => index !== indexToRemove);
  }

  return {
    ...toRefs(state),
    addToDo,
    removeToDo,
  };
}

type ToDoListStore = ReturnType<typeof toDoListStore>;

export const ToDoListKey: InjectionKey<ToDoListStore> = Symbol('ToDoListKey');

export function provideToDoListStore(store: ToDoListStore): void {
  provide(ToDoListKey, store);
}

export function useToDoListStore(): ToDoListStore {
  const store = inject(ToDoListKey, toDoListStore());
  return store;
}
