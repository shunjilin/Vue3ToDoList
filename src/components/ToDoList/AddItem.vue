<template>
  <input @keydown.enter="addToDo" type="text" v-model="addToDoInput" />
  <button @click="addToDo">Add</button>
</template>

<script lang="ts">
import { defineComponent } from 'vue';
import { useToDoListStore } from './store';
export default defineComponent({
  setup() {
    const { addToDo, addToDoInput } = useToDoListStore();

    return {
      addToDo,
      addToDoInput,
    };
  },
});
</script>

<style scoped>
button {
  background-color: #009688;
  color: white;
  margin: 0 0.4rem;
  padding: 0.4rem 1rem;
  border-radius: 0.2rem;
  border: none;
  cursor: pointer;
}
input {
  border: 1px solid #525858;
  padding: 0.2rem 0.4rem;
}
</style>
