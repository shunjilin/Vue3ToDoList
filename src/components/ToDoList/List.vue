<template>
  <ul>
    <li v-for="(item, index) in list" :key="item.created">
      <input type="checkbox" v-model="item.completed" />
      <span :class="item.completed && 'completed'">{{ item.title }}</span>
      <span class="close" @click="removeToDo(index)">Ã—</span>
    </li>
  </ul>
</template>

<script lang="ts">
import { defineComponent } from 'vue';
import { useToDoListStore } from './store';
export default defineComponent({
  setup() {
    const { list, removeToDo } = useToDoListStore();

    return {
      list,
      removeToDo,
    };
  },
});
</script>

<style scoped>
ul {
  list-style-type: none;
  padding: 0;
  margin: 25px auto;
  text-align: left;
  vertical-align: middle;
}
li {
  border: 1px solid#525858;
  padding: 0.2rem 0.5rem;
  border-top: none;
  display: flex;
}
li:first-child {
  border-top: 1px solid #525858;
}
li > * + * {
  margin-left: 0.4rem;
}
.close {
  cursor: pointer;
  color: red;
  margin-left: auto;
}
.completed {
  text-decoration: line-through;
}
</style>
