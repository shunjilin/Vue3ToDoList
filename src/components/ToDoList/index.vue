<template>
  <div>
    <h1>ToDo List</h1>
    <AddItem />
    <List />
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue';
import { provideToDoListStore, useToDoListStore } from './store';
import List from './List.vue';
import AddItem from './AddItem.vue';

export default defineComponent({
  components: {
    List,
    AddItem,
  },
  setup() {
    const store = useToDoListStore();
    provideToDoListStore(store);
  },
});
</script>

<style scoped>
div {
  text-align: center;
  margin: 0 auto;
  max-width: 300px;
}
</style>
