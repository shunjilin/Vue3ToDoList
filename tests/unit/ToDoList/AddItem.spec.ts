import { shallowMount } from '@vue/test-utils';
import AddItem from '@/components/ToDoList/AddItem.vue';
import { ToDoListKey } from '@/components/ToDoList/store';

describe('AddItem', () => {
  it('renders', () => {
    const wrapper = shallowMount(AddItem, {
      global: {
        provide: {
          [ToDoListKey as any]: {},
        },
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('calls addToDo function on clicking add button', () => {
    const addToDo = jest.fn();
    const wrapper = shallowMount(AddItem, {
      global: {
        provide: {
          [ToDoListKey as any]: {
            addToDo,
          },
        },
      },
    });
    wrapper.get('button').trigger('click');
    expect(addToDo).toHaveBeenCalled();
  });

  it('calls addToDo function on enter keydown on input', () => {
    const addToDo = jest.fn();
    const wrapper = shallowMount(AddItem, {
      global: {
        provide: {
          [ToDoListKey as any]: {
            addToDo,
          },
        },
      },
    });
    wrapper.get('input').trigger('keydown.enter');
    expect(addToDo).toHaveBeenCalled();
  });
});
