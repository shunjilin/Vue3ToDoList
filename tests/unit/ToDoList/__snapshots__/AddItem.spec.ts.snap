// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddItem renders 1`] = `
<div
  data-v-app=""
  id="app"
>
  
  <input
    type="text"
  />
  <button>
    Add
  </button>
  
</div>
`;
